File: ios/Host/ChartPriceScaleChanges.swift
// Copyright 2026 Kirill Novikov
// SPDX-License-Identifier: MIT

struct ChartPriceScaleChange {
  let paneId: String
  let priceScaleId: String
  let scale: Double
  let isMainPane: Bool
}

struct ChartScaleYResult {
  let stateChanged: Bool
  let priceScaleChange: ChartPriceScaleChange?
}

struct ChartPriceScaleChanges {
  private var changes: [ChartPriceScaleChange] = []

  mutating func record(_ result: ChartScaleYResult) {
    guard let change = result.priceScaleChange else { return }
    if let index = changes.firstIndex(where: {
      $0.paneId == change.paneId && $0.priceScaleId == change.priceScaleId
    }) {
      changes[index] = change
    } else {
      changes.append(change)
    }
  }

  mutating func clear() {
    changes.removeAll(keepingCapacity: true)
  }

  mutating func emit(
    mainScale: (Double) -> Void,
    priceScale: (ChartPriceScaleChange) -> Void
  ) {
    let pending = changes
    clear()
    for change in pending {
      if change.isMainPane { mainScale(change.scale) }
      priceScale(change)
    }
  }
}

File: android/src/main/java/com/tradingcharts/ChartPriceScaleChanges.kt
package com.tradingcharts

internal data class ChartPriceScaleChange(
    val paneId: String,
    val priceScaleId: String,
    val scale: Double,
    val isMainPane: Boolean,
)

// Reused per view. These small output arrays carry ScaleYResult across JNI;
// they are independent of the snapshot's positional ABI.
internal class ChartScaleYResult {
  val numbers = DoubleArray(NUMBER_COUNT)
  val strings = arrayOf("", "")

  fun priceScaleChange(): ChartPriceScaleChange? =
      if (numbers[SCALE_CHANGED] == 0.0) null
      else
          ChartPriceScaleChange(
              strings[PANE_ID],
              strings[PRICE_SCALE_ID],
              numbers[SCALE],
              numbers[IS_MAIN_PANE] != 0.0,
          )

  companion object {
    const val SCALE_CHANGED = 0
    const val SCALE = 1
    const val IS_MAIN_PANE = 2
    const val NUMBER_COUNT = 3
    const val PANE_ID = 0
    const val PRICE_SCALE_ID = 1
  }
}

internal class ChartPriceScaleChanges {
  private val changes = mutableListOf<ChartPriceScaleChange>()

  val isPending: Boolean
    get() = changes.isNotEmpty()

  fun record(change: ChartPriceScaleChange?) {
    if (change == null) return
    val index = changes.indexOfFirst {
      it.paneId == change.paneId && it.priceScaleId == change.priceScaleId
    }
    if (index >= 0) changes[index] = change else changes.add(change)
  }

  fun clear() = changes.clear()

  fun emit(mainScale: (Double) -> Unit, priceScale: (ChartPriceScaleChange) -> Unit) {
    if (changes.isEmpty()) return
    val pending = changes.toList()
    clear()
    for (change in pending) {
      if (change.isMainPane) mainScale(change.scale)
      priceScale(change)
    }
  }
}
